src/main/java/com/google/collegeplanner/servlets/PlannerServlet.java

```diff
@@ -29,35 +35,221 @@
 /** Servlet that returns the mutli-semester plan for the given courses.*/
 @WebServlet("/api/planner")
 public class PlannerServlet extends HttpServlet {
+  private HashMap<String, Integer> indegree; // Number of prerequisites for a course
+  private HashMap<String, HashSet<String>>
+      nextCourses; // Courses that come after the given course in the graph
+  private HashMap<String, HashSet<String>> corequisites;
+  private HashMap<String, Integer> depths; // depth of a course in the graph
+  private HashMap<String, Integer> credits; // Number of credits for each course
+  private ArrayList<String> courseList;
+  private int totalCredits;
+
   /**
-   * Organizes courses into the given number of semesters
+   * Organizes courses from POST request into a given number of semesters
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // TODO(ramyabuva): Implement algorithm to split into the number of semesters
     JSONObject body;
+    int numSemesters;
+    JSONArray selectedClasses;
     try {
       body = getBody(request);
-    } catch (ParseException | NullPointerException e) {
+      numSemesters = Integer.parseInt((String) body.get("semesters"));
+      selectedClasses = (JSONArray) body.get("selectedClasses");
+    } catch (Throwable e) {
       respondWithError(
           "Invalid body for POST request.", HttpServletResponse.SC_BAD_REQUEST, response);
       return;
     }
-    String strClasses = (String) body.get("selectedClasses");
-    // Returns the courses in a single semester
-    String[][] courses = {strClasses.split(",")};
+
+    constructGraphs(selectedClasses);
+
+    ArrayList<ArrayList<String>> semesters = getPlan(numSemesters);
+    // If courseList is not empty, scheduling was not possible
+    if (!courseList.isEmpty()) {
+      return;
+    }
     response.setContentType("applications/json;");
-    response.getWriter().println(new Gson().toJson(courses));
+    response.getWriter().println(new Gson().toJson(semesters));
   }
 
-  private static JSONObject getBody(HttpServletRequest request)
+  /**
+   * Creates and returns a valid plan for the given number of semesters
+   *
+   * @param numSemesters The number of semesters remaining
+   */
+  private ArrayList<ArrayList<String>> getPlan(int numSemesters) {
+    ArrayList<ArrayList<String>> semesters = new ArrayList<>();
+    for (int i = numSemesters; i > 0; i--) {
+      int avgCredits = totalCredits / i;
+      ArrayList<String> semester = new ArrayList<>();
+      for (int j = 0; j < courseList.size(); j++) {
+        String course = courseList.get(j);
+        if (indegree.get(course) == 0) {
+          // Check if all corequisites also have an indegree of 0
+          if (isValidToAdd(course)) {
+            // If valid, add all to semester
+            courseList.remove(j);
+            j--;
```

Comment:
I considered this, but I realized I would still be maintaining a variable to iterate through the list, and you don't want to repeat until the courseList is empty, but some other conditions. Also, since j begins at 0 and is never less than 0 when the condition is evaluated, this for loop should never produce an error.